### ISSUE TITLE
Spike: Research queue and broker alternatives

### ISSUE THREAD

**Issue by katherine-black** [cloudigrade/cloudigrade#528]
In GitLab by @infinitewarp on Jan 22, 2019, 15:55

Consider alternatives to Celery and SQS.

Why are we doing this? We have bizogus errors when working with SQS via Celery + Kombu + pycurl. SQS also doesn't play well as a true queue. Also SQS randomly gives boguzarro errors. [citation needed]

Maybe only replace SQS? Maybe replace both?

Expected output:
- Some kind of prototype using at least one non-SQS queue and task system.
- Prototype should have decorators, ability to schedule, and ability to have multiple workers.
- Prototype task functions should be able to be called both async and sync.
- Prototype should have regular queues configurable with DLQ fail-overs after X-configured attempts/reads.

Alternatives to consider:
- [Celery](http://docs.celeryproject.org/en/latest/index.html) with [RabbitMQ](https://www.rabbitmq.com/) or [Redis](https://redis.io/)
- [RQ](http://python-rq.org) with [Redis](https://redis.io/) only
- [dramatiq](https://github.com/Bogdanp/dramatiq) with [RabbitMQ](https://www.rabbitmq.com/) or [Redis](https://redis.io/)
- [Django Q](https://github.com/Koed00/django-q/) with [RabbitMQ](https://www.rabbitmq.com/) or [Redis](https://redis.io/)

Two caveats:
- SQS must remain for receiving notifications from S3 bucket events.
  - Or is there a better way?
- We can't run our own queue INSIDE OF the insights cluster. However, we could run it in AWS! Or maybe GCP. Maybe.
  - Because Insights (really OSD) only allows us to expose HTTP ports 80 and 443.

Time box:
- 1 calendar week.

**Comment by katherine-black** [comment]
In GitLab by @katherine-black on Jan 22, 2019, 15:56

https://github.com/Bogdanp/dramatiq

**Comment by katherine-black** [comment]
In GitLab by @katherine-black on Jan 22, 2019, 15:56

http://python-rq.org

**Comment by katherine-black** [comment]
In GitLab by @werwty on Jan 29, 2019, 15:34

Pulp switched from rabbitmq to rq last year.
Here's the write up on it: https://pulpproject.org/2018/05/08/pulp3-moving-to-rq/

And this change was discussed in great detail on the pulp mailing list:
https://www.redhat.com/archives/pulp-dev/2018-March/msg00101.html

**Comment by katherine-black** [comment]
In GitLab by @infinitewarp on Feb 4, 2019, 13:56

I've been experimenting with RQ, and it may be too bare-bones and not suitable for our needs.

RQ requires additional packages for Django (`django-rq`) and scheduling (`rq-scheduler`). That scheduler library appears to be intended for scheduling one-off jobs in the future, but not for repeating jobs like `cron`. Some one-off solutions exist in StackOverflow [(here)](https://stackoverflow.com/a/18595392) and GitHub [(here)](https://github.com/rq/rq-scheduler/issues/51#issuecomment-362352497) comment that attempt to work around this limitation.

RQ has a built-in "failed queue" (its take on a dead-letter queue), but there is no standard support for retrying. A failed job _immediately_ migrates to the failed queue. I found a library [rq_retry_scheduler](https://github.com/mikemill/rq_retry_scheduler) that provides a [Worker](https://github.com/mikemill/rq_retry_scheduler/blob/master/rq_retry_scheduler/worker.py) class that includes an exception handler for retrying failed jobs before moving it to the failed queue, but... that library hasn't been updated in two years, it has no watchers/stars/forks on GitHub (which suggests nobody uses it), and the aforementioned Worker class has baked-in back-off without a clear way of changing or configuring it (either globally or at a per-job-function level. That could provide inspiration for us writing our own worker class, but that's more infrastructure for us maintain.

RQ appears to only provide an abstraction over Redis for _job_ queues, not plain FIFO object/data queues. So, we would need to write our own helpers to interact directly with Redis. The functionality of Redis that libraries like RQ use functions like [RPOPLPUSH](https://redis.io/commands/rpoplpush) to interact with the head and tail of [lists](https://redis.io/topics/data-types#lists) ([more list commands](https://redis.io/commands#list)). Redis doesn't have "queues" as a first-class type of object.

**Comment by katherine-black** [comment]
In GitLab by @infinitewarp on Feb 20, 2019, 13:45

Some final notes before wrapping up this spike…

Dramatiq seems like a promising candidate, but it requires a third-party package for scheduled/recurring tasks, and it requires another third-party package for Django integration. I have pushed a partially-complete branch https://gitlab.com/cloudigrade/cloudigrade/tree/dramatiq, but the Django integration piece doesn't seem to work with out env-based configuration. I feel like Dramatiq might fit much better with non-Django Python projects.

RabbitMQ still seems like a good idea, especially since Celery supports it "out of the box", but I struggled getting it to work in a good cross-environment setup. Celery could run in an AWS EC2 instance, but there is no officially-supported Celery AMI. I started with a recent CentOS AMI and manually installed Celery, and that works fine. However, the next trick is securing communication in and out of that EC2 instance. RabbitMQ has TLS support with custom certs, and I was able to get that working on the server, but I couldn't find a way to make Celery use the custom certs and communicate with TLS as a client.

An alternative would be to run Celery in OSD and use it _only_ internally with the cloudigrade task workers. This gets us the benefit of greater reliability, but it adds complexity by leaving SQS for the traditional queues used by houndigrade.

So, I'm leaving things largely unchanged at the end of this spike. I have created one MR in houndigrade that allows us to use the default Celery queue in the future (we _should have_ always included this) if we want to simplify the configuration of new tasks.
